### ISSUE TITLE
export = locale in index.d.ts doesnt allow synthetic import

### ISSUE THREAD

**Issue by Alex33400** [iamkun/dayjs#2189]
**Describe the bug**
When I want to import dayjs as synthetic import in my lib, I get the following error : 

export = dayjs;
      ~~~~~~~~~~~~~~~
    This module is declared with using 'export =', and can only be used with a default import when using the 'allowSyntheticDefaultImports' flag

**Expected behavior**
Application can build with this line : "import dayjs from 'dayjs';"

**Information**
 - Day.js Version [e.g. v1.11.7]
 - OS: [Windows]
 - Browser: Chrome 108
 - Time zone: GMT +1

**Comment by BePo65** [comment]
As a starter: this does not only affect the 'core' program of dayjs (as handled in your pr #2190), but any plugin and locale file.

From your pr I suppose you are using dayjs in an esm environment. But dayjs is a commonjs library (the esm first implementation is still under construction; mostly called "dayjs 2.0", but according to the authors it will probably get 3.0).

I made a pr for fixing all the (default) type definitions for esm (pr #1581) and the answer was "let's implement it in dayjs 2.0" (at least that is my interpretation 😄).

The typescript [handbook ](https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require)says that `export =` in the file ZipCodeValidator requires an `import zip = require("./ZipCodeValidator");` in the module using it.
So perhaps that is an acceptable workaround until we get "dayjs 2.0".

**Comment by trim21** [comment]
> From your pr I suppose you are using dayjs in an esm environment. But dayjs is a commonjs library

But it also break esm build, you can't `import dayjs from 'dayjs/esm'` without `allowSyntheticDefaultImports` flag.

**Comment by BePo65** [comment]
How about using `import zip = require("dayjs")` as the handbook says instead of `import dayjs from 'dayjs/esm`?

Or `import * as dayjs from 'dayjs'` as the dayjs [documentation](https://day.js.org/docs/en/installation/typescript) recomments?

In my (esm) application, I gave up to try to use 'dayjs/esm' because of the effects you see in your app too.

**Comment by trim21** [comment]
TBH, `import * as dayjs from 'dayjs'` is not semantic correct.

This means it import all named exports and put them as properties of `dayjs`. It means `dayjs` here is a object, it's not callable.

and `import dayjs = require("dayjs");` actually only works in `.d.ts`, but not `.ts`, you can't compile it with tsc.

```text
main.ts:9:1 - error TS1202: Import assignment cannot be used when targeting ECMAScript modules. Consider using 'import * as ns from "mod"', 'import {a} from "mod"', 'import d from "mod"', or another module format instead.

9 import dayjs = require("dayjs");
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 1 error in main.ts:9
```

**Comment by BePo65** [comment]
You are right: `import * as ...` is for commonjs environments and not usable in your case.

It is quite a while since i wrote pr #1581, but as far as I remember the problems were:
* the type definitions included in dayjs are for use in commonjs environments
* during the build process a directory 'dayjs/esm' with the esm version of dayjs is generated.
* unluckily the build script just copies the type definitions from dayjs (the commonjs version) to this directory
* in my pr I tried to fix this by introducing a second set of type definitions that handle especially the `export default` problem. But the author did not like my idea of 2 sets of type definitions (I can accept this attitude) and I did not find a way to get away with 1 common type definition (AFAIK `export default ..` prevents the use in commonjs environments).

So perhaps a (temporary) solution could be to use a copy of the type definitions from my pr in the local "typeRoots" of your program.

**Comment by trim21** [comment]
> It is quite a while since i wrote pr #1581, but as far as I remember the problems were:

I read your pr, the solution is very simple...

there is no need move type components defs, just add a `esm/index.d.ts` file

```ts
/// <reference path="./locale/index.d.ts" />
import dayjs = require('dayjs')
export default dayjs;
```

**Comment by BePo65** [comment]
I think I still have my test project from the time of this pr. If I find it, I could try it in a commonjs environment (but not in the next 2 minutes 😄).

**Comment by trim21** [comment]
On second thought I think it could be a fix... and it doesn't touch code with commonjs.

**Comment by BePo65** [comment]
ASAIK type definitions are for typescript and typescript can be used for esm and commonjs environments. So we need type definitions for both systems and regarding `export` these are different. So we need 2 different type definitins.

Or did I miss something.

**Comment by trim21** [comment]
> ASAIK type definitions are for typescript and typescript can be used for esm and commonjs environments. So we need type definitions for both systems and regarding `export` these are different. So we need 2 different type definitins.
> 
> Or did I miss something.

because `export default` also works with cjs. you don't have to write `export = dayjs`

**Comment by iamkun** [comment]
Thanks for trying to figure out this historic problem.

There are some problems like https://github.com/iamkun/dayjs/issues/277 and https://github.com/iamkun/dayjs/issues/313, which do not have a proper fix in the current project design. (At least to me) Most of them are related to Typescript and ESM, which is essential at present.

Maybe a 2.0 version will be a better place to fix it and get rid of all the limitations on the current code.

**Comment by trim21** [comment]
> Thanks for trying to figure out this historic problem.
> 
> There are some problems like #277 and #313, which do not have a proper fix in the current project design. (At least to me) Most of them are related to Typescript and ESM, which is essential at present.
> 
> Maybe a 2.0 version will be a better place to fix it and get rid of all the limitations on the current code.

I think https://github.com/iamkun/dayjs/pull/2193 is the fix, could you take it look? it doesn't touch cjs files, and only change some code in `dayjs/esm`.

**Comment by iamkun** [comment]
True, the plugin system introduced in dayjs is not really flowed the TypeScript way.  That's maybe the main challenge in 2.0 version.

**Comment by trim21** [comment]
Current interface merging is already the typescript way to support plugin IMO, it's what typescript do in their document. 🤔

https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation

**Comment by BePo65** [comment]
That is the easy part; problems arise, when you try to add a setter to a getter defined in core via interface merging. And you have to "preview" all the expansion points needed by plugins(even plugins that you don't know of 😄).

**Comment by trim21** [comment]
> when you try to add a setter to a getter defined in core via interface merging

Can I get a example?

**Comment by BePo65** [comment]
See pr #2128 (but it is a quite extensive pr; an implementation of the utc plugin for dayjs 2.0).

In dayjs "core" we have a "getter" named `utcOffset(): number` and the plugin utc should add a setter for it (`utcOffset(offset: number | string, keepLocalTime?: boolean): Dayjs`).

The problem is not the js code (the way dayjs 1,x does it, will of course also run in dayjs 2.0 - but not in a typesafe way), I tried several ways to solve it, using my own ideas and any "trick" I found searching stackoverflow and other resources. But I could not define the overload signature in the utc module in a way that the compiler does not show an error **and** that the code does not throw during runtime.

In the end I had to add the overload signature to the "core" module - but IMO a plugin should not require a modification of the module to be expanded (but it is hard to identify the "expansion" points beforehand).

Any better solutions are welcome 😄 .
